dali/pipeline/operators/util/external_source.h

```diff
@@ -73,34 +124,34 @@ class ExternalSource : public Operator<Backend> {
     // Note: If we create a GPU source, we will need to figure
     // out what stream we want to do this copy in. CPU we can
     // pass anything as it is ignored.
+    std::unique_lock<std::mutex> busy_lock(busy_m_);
+    auto data = t_data_.GetEmpty();
+    auto is_data_tl = data_in_tl_.GetEmpty();
+    busy_lock.unlock();
 
-    std::unique_lock<std::mutex> lock(busy_m_);
-    cv_.wait(lock,  [this]{return !this->busy_;});
-
-    t_data_.resize(t.size());
+    data->resize(t.size());
     for (size_t i = 0; i < t.size(); ++i) {
-      t_data_[i].Copy(t[i], 0);
+      (*data)[i].Copy(t[i], 0);
     }
-    data_in_tl_ = false;
-    busy_ = true;
+    *is_data_tl = false;
+    busy_lock.lock();
+    t_data_.AddBack(data);
+    data_in_tl_.AddBack(is_data_tl);
+    cv_.notify_one();
   }
 
   DISABLE_COPY_MOVE_ASSIGN(ExternalSource);
 
  protected:
-  void RunImpl(Workspace<Backend> *ws) override;
+  void RunImpl(WorkspaceBatch<Backend> *ws) override;
 
   string output_name_;
-  TensorList<Backend> tl_data_;
-  std::vector<Tensor<Backend>> t_data_;
-  bool data_in_tl_;
+  detail::CachingList<TensorList<Backend>> tl_data_;
+  detail::CachingList<std::vector<Tensor<Backend>>> t_data_;
+  detail::CachingList<bool> data_in_tl_;
```

Comment:
I want to avoid creating and freeing TenosrList memory. And as far as I know I cannot detach element from std::list, do something, then move it to another one. Vector could do but adding things at the front is rather inefficient.

Reply:
Oh, I meant the list of unique_ptrs to bool.

Reply:
Your are right. Done.